specify/specify7
# Do not set businessrule saveblockers sequentially

There is currently a concerning bug on the `xml-editor` branch where the save blockers for business rules are not being shown. 

On  `xml-editor`: 

https://github.com/specify/specify7/assets/64045831/fbb8104b-4bfa-422a-a3fa-8206ce4f204d

### Dev Notes

The order of business rules whenever a field is changed is the following: 
- fieldChecks from the table's [business rule definitions](https://github.com/specify/specify7/blob/production/specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts)
- uniqueness rules
- tree-related business rules (if applicable)

https://github.com/specify/specify7/blob/1c85f365add6795ae841f892cfe9f6d2c30802d7/specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts#L114-L123

Each of these checks return a BusinessRuleResult. 
The problem here is that the businessrule saveblockers were being set one at a time, by iterating over the `checks` array:

https://github.com/specify/specify7/blob/1c85f365add6795ae841f892cfe9f6d2c30802d7/specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts#L148-L152

This does not seem to be the intended way of setting save blockers, which expects an array of strings corresponding to every save blocker message which each get converted to a `blocker`. 

The following snippet is the most relevant from setSaveBlockers()
 https://github.com/specify/specify7/blob/1c85f365add6795ae841f892cfe9f6d2c30802d7/specifyweb/frontend/js_src/lib/components/DataModel/saveBlockers.tsx#L108-L111

Consider the case in the video provided above. 
`checks` would become an array with the following:
- CollectionObject has no catalogNumber fieldCheck -> so this is valid
- There _is_ another CO with the same catalognumber -> invalid
- CO is not a tree resource -> valid

When the uniqueness rule save blocker is set, it then gets to the tree business rule checks which is valid. 
Due to the way setSaveBlockers() is implemented, the uniqueness rule save blockers become 'overridden' and filtered out, leaving only the valid tree business rule check. 

@maxxxxxdlp is this behavior intentional? Practically, it would imply only one 'type' of saveBlockers would be shown at a time. 
For example consider the case where there is both a parsing error and a business rule violation for a field. The error shown to the user will be the one _last_ set. 
This could lead to another bug as well: if parsing saveBlockers are applied first followed by businessrules, consider the case where the parsing is invalid but businessrules is valid (not that likely in the wild, however). 
The field would theoretically be OK to save because the parsing blockers would be filtered out. 

- - - 

### Testing Instructions
- Make sure the three types of uniqueness rules are properly indicated to the user 
  - Examples are provided in https://github.com/specify/specify7/pull/3851#issue-1824735764
  - A list of every uniqueness rule is provided in https://github.com/specify/specify7/pull/3996#issue-1888914474 or on the [Speciforum](https://discourse.specifysoftware.org/t/business-rules-and-logic-in-specify-7/1231)
